cloud/docker/docker_login.py

```diff
@@ -0,0 +1,246 @@
+        if self.has_changed() and not self.module.check_mode:
+            self.update_dockercfg()
+
+    # This is what the underlaying docker-py unfortunately doesn't do (yet).
+    def update_dockercfg(self):
+
+        # Create dockercfg file if it does not exist.
+        if not os.path.exists(self.dockercfg_path):
+            open(self.dockercfg_path, "w")
+            self.log.append("Created new Docker config file at %s" % self.dockercfg_path)
+        else:
+            self.log.append("Updated existing Docker config file at %s" % self.dockercfg_path)
+
+        # Get existing dockercfg into a dict.
+        try:
+            docker_config = json.load(open(self.dockercfg_path, "r"))
+        except ValueError:
+            docker_config = dict()
+        if not docker_config.has_key(self.registry):
+            docker_config[self.registry] = dict()
+        docker_config[self.registry] = dict(
+            auth  = base64.b64encode(self.username + b':' + self.password),
+            email = self.email
+        )
+
+        # Write updated dockercfg to dockercfg file.
+        try:
+            json.dump(docker_config, open(self.dockercfg_path, "w"), indent=4, sort_keys=True)
+        except Exception as e:
+            self.module.fail_json(msg="failed to write auth details to file", error=repr(e))
+
+    # Compatible to docker-py auth.decode_docker_auth()
+    def encode_docker_auth(self, auth):
+        s = base64.b64decode(auth)
+        login, pwd = s.split(b':', 1)
+        return login.decode('ascii'), pwd.decode('ascii')
+
+    def get_msg(self):
+        return ". ".join(self.log)
+
+    def has_changed(self):
+        return self.changed
+
+def main():
+
+    module = AnsibleModule(
+        argument_spec = dict(
+            registry        = dict(required=True),
+            username        = dict(required=True),
+            password        = dict(required=True),
+            email           = dict(required=False, default='anonymous@localhost.local'),
+            reauth          = dict(required=False, default=False, type='bool'),
+            dockercfg_path  = dict(required=False, default='~/.dockercfg'),
+            docker_url      = dict(default='unix://var/run/docker.sock'),
+            timeout         = dict(default=10, type='int')
+        ),
+        supports_check_mode=True
+    )
+
+    if not has_lib_docker:
+        module.fail_json(msg="python library docker-py required: pip install docker-py==1.1.0")
+
+    if not has_lib_requests_execeptions:
+        module.fail_json(msg="python library requests required: pip install requests")
+
+    try:
+        manager = DockerLoginManager(module)
+        manager.login()
+        module.exit_json(changed=manager.has_changed(), msg=manager.get_msg())
+
+    except Exception as e:
+        module.fail_json(msg="Module execution has failed due to an unexpected error", error=repr(e))
+
+# import module snippets
+from ansible.module_utils.basic import *
+
+main()
```

Comment:
Stylistically it's nice to use this:
``` python
if __name__ == '__main__':
    main()
```

That way if we ever need unittests for this module we can import it without running the module.